### PR TITLE
core[minor]docs[minor]: new `tool` wrapper

### PR DESCRIPTION
**Pull request by bracesproul**

### CODE REVIEW

File: docs/core_docs/docs/how_to/custom_tools.ipynb [hunk 1]
```diff
@@ -26,6 +26,55 @@
...
+   "metadata": {},
+   "source": [
+    "## `tool` function\n",
+    "\n",
+    ":::note\n",
+    "Only available in `@langchain/core` version 0.2.X and above.\n",
+    ":::\n",
+    "\n",
+    "\n",
+    "The [`tool`](https://v02.api.js.langchain.com/classes/langchain_core_tools.tool.html) function is used to wrap a generic JavaScript/TypeScript function, along with additional arguments that define your tool. The `tool` function will return an instance of the [`StructuredTool`](https://v02.api.js.langchain.com/classes/langchain_core_tools.StructuredTool.html) class, so it is compatible with all the existing tool calling infrastructure in the LangChain library."
```
Comment (jacoblee93) [inline docs/core_docs/docs/how_to/custom_tools.ipynb:41]: nit: Remove the `v02` to make it easier to update later

File: docs/core_docs/docs/how_to/custom_tools.ipynb [hunk 2]
```diff
@@ -26,6 +26,55 @@
...
   },
+  {
+   "cell_type": "markdown",
+   "id": "f6ec6ee8",
+   "metadata": {},
+   "source": [
+    "## `tool` function\n",
+    "\n",
+    ":::note\n",
+    "Only available in `@langchain/core` version 0.2.X and above.\n",
```
Comment (jacoblee93) [inline docs/core_docs/docs/how_to/custom_tools.ipynb:37]: nit: have been using lower case `x` to match `npm` style:

https://www.npmjs.com/package/semver

File: docs/core_docs/docs/how_to/custom_tools.ipynb [hunk 3]
```diff
@@ -26,6 +26,55 @@
...
+   ],
+   "source": [
+    "import { z } from \"zod\";\n",
+    "import { tool } from \"@langchain/core/tools\";\n",
+    "\n",
+    "const adderSchema = z.object({\n",
+    "  a: z.number(),\n",
+    "  b: z.number(),\n",
+    "});\n",
+    "const adderTool = tool((input: z.infer<typeof adderSchema>): Promise<string> => {\n",
```
Comment (jacoblee93) [inline docs/core_docs/docs/how_to/custom_tools.ipynb:66]: Do we need the type annotation here? Should be inferred?

File: langchain-core/src/messages/base.ts [hunk 4]
```diff
@@ -53,7 +53,7 @@ export type MessageContentComplex =
   | MessageContentText
   | MessageContentImageUrl
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  | (Record<string, any> & { type?: "text" | "image_url" | string })
+  | (Record<string, any> & { type?: "text" | "image_url" | "tool" | string })
```
Comment (jacoblee93) [inline langchain-core/src/messages/base.ts:56]: Don't think we should be touching this - it's not in our message spec

File: langchain-core/src/tools.ts [hunk 5]
```diff
@@ -33,7 +36,7 @@ export class ToolInputParsingException extends Error {
 
 export interface StructuredToolInterface<
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  T extends z.ZodObject<any, any, any, any> = z.ZodObject<any, any, any, any>
+  T extends ZodAny = ZodAny
```
Comment (jacoblee93) [inline langchain-core/src/tools.ts:39]: Nice, but remove unnecessary `// eslint-disable-next-line @typescript-eslint/no-explicit-any`

File: langchain-core/src/tracers/base.ts [hunk 6]
```diff
@@ -392,7 +392,11 @@ export abstract class BaseTracer extends BaseCallbackHandler {
     return run;
   }
 
-  async handleToolEnd(output: string, runId: string): Promise<Run> {
+  async handleToolEnd(
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    output: string | Record<string, any>,
```
Comment (jacoblee93) [inline langchain-core/src/tracers/base.ts:397]: Thought we weren't doing this yet?

File: docs/core_docs/docs/how_to/custom_tools.ipynb (hunk 2 above)
Comment (bracesproul) [reply docs/core_docs/docs/how_to/custom_tools.ipynb:37]: Planning on replacing this before merge when I know the actual version number it'll be

File: langchain-core/src/tracers/base.ts (hunk 6 above)
Comment (bracesproul) [reply langchain-core/src/tracers/base.ts:397]: whoops, meant to remove when I merged branches. thanks

File: langchain-core/src/tools.ts [hunk 7]
```diff
@@ -364,3 +367,60 @@ export abstract class BaseToolkit {
...
+    z.object({ input: z.string().optional() }).transform((obj) => obj.input);
+
+  const description =
+    fields.description ?? schema.description ?? `${fields.name} tool`;
+  return new DynamicStructuredTool<RunInput>({
+    name: fields.name,
+    description,
+    schema: schema as RunInput,
+    func: (input, _runManager, config) => {
+      return Promise.resolve(func(input, config));
```
Comment (jacoblee93) [inline langchain-core/src/tools.ts:423]: Think this can just be an `async` function no?

**Comment by jacoblee93** [comment]
Nice! Excited for this!

**Pull request merged**